File: functions/src/services/calendar.ts
/**
 * Calendar service — swappable behind an interface.
 *
 * Auto-selects at first use: if CALENDAR_SERVICE_ACCOUNT_JSON is set, a real
 * Google Calendar implementation is used (service account with domain-wide
 * delegation, impersonating each target member); otherwise a logging stub. So
 * the emulator and any unconfigured environment behave exactly as before.
 *
 * Setup (admin): see README "Google Calendar — domain-wide delegation". Grant the
 * service account the calendar + calendar.events scopes, then set
 * CALENDAR_SERVICE_ACCOUNT_JSON (path or inline JSON) in the functions env.
 */

import * as fs from "fs";
import { Timestamp } from "firebase-admin/firestore";
import { google, calendar_v3 } from "googleapis";

const TZ = "Asia/Kolkata";
const SCOPES = [
  "https://www.googleapis.com/auth/calendar",
  "https://www.googleapis.com/auth/calendar.events",
];

export interface CalendarService {
  isFree(email: string, start: Timestamp, end: Timestamp): Promise<boolean>;
  createHold(opts: {
    email: string;
    title: string;
    start: Timestamp;
    end: Timestamp;
    description?: string;
  }): Promise<void>;
  createReminder(opts: {
    email: string;
    title: string;
    due: Timestamp;
    description?: string;
  }): Promise<void>;
}

class StubCalendarService implements CalendarService {
  async isFree(): Promise<boolean> {
    return true;
  }
  async createHold(opts: { email: string; title: string }): Promise<void> {
    console.log(`[calendar:STUB] hold -> ${opts.email} | ${opts.title}`);
  }
  async createReminder(opts: { email: string; title: string }): Promise<void> {
    console.log(`[calendar:STUB] reminder -> ${opts.email} | ${opts.title}`);
  }
}

interface SaCreds {
  client_email: string;
  private_key: string;
}

function loadCreds(envVal: string): SaCreds {
  const trimmed = envVal.trim();
  const raw = trimmed.startsWith("{") ? trimmed : fs.readFileSync(trimmed, "utf8");
  const json = JSON.parse(raw);
  return { client_email: json.client_email, private_key: json.private_key };
}

class GoogleCalendarService implements CalendarService {
  constructor(private readonly creds: SaCreds) {}

  private clientFor(userEmail: string): calendar_v3.Calendar {
    const auth = new google.auth.JWT({
      email: this.creds.client_email,
      key: this.creds.private_key,
      scopes: SCOPES,
      subject: userEmail, // impersonate the member (domain-wide delegation)
    });
    return google.calendar({ version: "v3", auth });
  }

  async isFree(email: string, start: Timestamp, end: Timestamp): Promise<boolean> {
    try {
      const cal = this.clientFor(email);
      const res = await cal.freebusy.query({
        requestBody: {
          timeMin: start.toDate().toISOString(),
          timeMax: end.toDate().toISOString(),
          items: [{ id: "primary" }],
        },
      });
      const busy = res.data.calendars?.["primary"]?.busy || [];
      return busy.length === 0;
    } catch (err) {
      // Fail-open: don't block assignment because of a calendar error.
      console.error(`[calendar] isFree(${email}) failed:`, (err as Error).message);
      return true;
    }
  }

  async createHold(opts: {
    email: string;
    title: string;
    start: Timestamp;
    end: Timestamp;
    description?: string;
  }): Promise<void> {
    try {
      await this.clientFor(opts.email).events.insert({
        calendarId: "primary",
        requestBody: {
          summary: opts.title,
          description: opts.description || "",
          start: { dateTime: opts.start.toDate().toISOString(), timeZone: TZ },
          end: { dateTime: opts.end.toDate().toISOString(), timeZone: TZ },
        },
      });
    } catch (err) {
      console.error(`[calendar] createHold(${opts.email}) failed:`, (err as Error).message);
    }
  }

  async createReminder(opts: {
    email: string;
    title: string;
    due: Timestamp;
    description?: string;
  }): Promise<void> {
    try {
      const due = opts.due.toDate();
      const end = new Date(due.getTime() + 30 * 60000);
      await this.clientFor(opts.email).events.insert({
        calendarId: "primary",
        requestBody: {
          summary: opts.title,
          description: opts.description || "",
          start: { dateTime: due.toISOString(), timeZone: TZ },
          end: { dateTime: end.toISOString(), timeZone: TZ },
        },
      });
    } catch (err) {
      console.error(`[calendar] createReminder(${opts.email}) failed:`, (err as Error).message);
    }
  }
}

let cached: CalendarService | null = null;

function service(): CalendarService {
  if (cached) return cached;
  const credsEnv = process.env.CALENDAR_SERVICE_ACCOUNT_JSON;
  if (credsEnv) {
    try {
      cached = new GoogleCalendarService(loadCreds(credsEnv));
      console.info("[calendar] using Google Calendar");
    } catch (err) {
      console.error("[calendar] config error, falling back to stub:", (err as Error).message);
      cached = new StubCalendarService();
    }
  } else {
    cached = new StubCalendarService();
  }
  return cached;
}

export const calendarService: CalendarService = {
  isFree: (email, start, end) => service().isFree(email, start, end),
  createHold: (opts) => service().createHold(opts),
  createReminder: (opts) => service().createReminder(opts),
};

File: portal/services/calendar.py
"""
Google Calendar integration (docs/PRD.md §5.12).

Per-member OAuth consent, not a service account: IT declined the Workspace
domain-wide delegation a service account would need, so instead each team
member grants Calendar access themselves through the same Google sign-in the
portal already uses (see `SOCIALACCOUNT_PROVIDERS` in settings.py). Their
token is persisted by django-allauth (`SOCIALACCOUNT_STORE_TOKENS`) and looked
up here by email — every call site in this codebase checks or writes a
*different* member's calendar than whoever is logged in (auto-assignment,
coordinator reassignment, the deadline cron), so there is no "current user"
token to reuse.

Two fallback layers, both intentional, both ported from the old design:

* **Auto-fallback.** With `CALENDAR_ENABLED` off, every call becomes a logged
  no-op — dev/test runs the same code paths as production without hitting
  Google.
* **Fail open, per member.** A member who hasn't (re)consented yet, or whose
  token fails to refresh (e.g. revoked from
  myaccount.google.com/permissions), is treated as free/no-op for *them
  only* — one person's missing consent must never block the whole assignment
  pipeline. A flaky Google API must never be the reason an event goes
  unstaffed — a double-booked photographer is a smaller problem than no
  photographer.
"""

from __future__ import annotations

import logging
from datetime import datetime, timedelta

from django.conf import settings

logger = logging.getLogger(__name__)

TIMEZONE = "Asia/Kolkata"
SCOPES = [
    "https://www.googleapis.com/auth/calendar",
    "https://www.googleapis.com/auth/calendar.events",
]

#: How long a deadline reminder occupies the assignee's calendar.
REMINDER_MINUTES = 30


class StubCalendar:
    """Used whenever CALENDAR_ENABLED is off. Everything is free."""

    configured = False

    def is_free(self, email: str, start: datetime, end: datetime) -> bool:
        return True

    def create_hold(self, email: str, title: str, start: datetime, end: datetime, description: str = "") -> None:
        logger.info("[calendar:stub] hold -> %s | %s", email, title)

    def create_reminder(self, email: str, title: str, due: datetime, description: str = "") -> None:
        logger.info("[calendar:stub] reminder -> %s | %s", email, title)


class GoogleCalendar:
    """
    Real implementation: looks up each member's own stored Google OAuth token
    (granted at their login) and acts on their calendar with it — no service
    account, no impersonation.
    """

    configured = True

    def _credentials_for(self, email: str):
        """
        The member's live `Credentials`, or `None` if they haven't granted
        Calendar access yet (no stored refresh token) — callers treat that as
        the fail-open case, not an error.
        """
        from allauth.socialaccount.models import SocialAccount, SocialToken
        from google.auth.transport.requests import Request
        from google.oauth2.credentials import Credentials

        try:
            account = SocialAccount.objects.get(user__email__iexact=email, provider="google")
            token = SocialToken.objects.get(account=account)
        except (SocialAccount.DoesNotExist, SocialToken.DoesNotExist):
            return None
        if not token.token_secret:
            # No refresh token stored yet -- they signed in before the
            # calendar scope existed, or haven't logged in since
            # CALENDAR_ENABLED was turned on. They'll get one next time they
            # sign in (access_type=offline + prompt=consent, see settings.py).
            return None

        app = settings.SOCIALACCOUNT_PROVIDERS["google"]["APP"]
        creds = Credentials(
            token=token.token,
            refresh_token=token.token_secret,
            token_uri="https://oauth2.googleapis.com/token",
            client_id=app["client_id"],
            client_secret=app["secret"],
            scopes=SCOPES,
        )
        # Always refresh rather than trusting locally-cached expiry: it's one
        # cheap round trip, and it self-heals a stale/incorrect `expires_at`
        # instead of surfacing a 401 to the caller. A RefreshError (token
        # revoked) propagates up to is_free/_insert's own try/except, which
        # already fail open.
        creds.refresh(Request())
        token.token = creds.token
        token.expires_at = creds.expiry
        token.save(update_fields=["token", "expires_at"])
        return creds

    def _service(self, user_email: str):
        from googleapiclient.discovery import build

        creds = self._credentials_for(user_email)
        if creds is None:
            return None
        return build("calendar", "v3", credentials=creds, cache_discovery=False)

    def is_free(self, email: str, start: datetime, end: datetime) -> bool:
        try:
            service = self._service(email)
            if service is None:
                logger.info("calendar: %s hasn't granted Calendar access yet, treating as free", email)
                return True
            response = (
                service.freebusy()
                .query(
                    body={
                        "timeMin": start.isoformat(),
                        "timeMax": end.isoformat(),
                        "items": [{"id": "primary"}],
                    }
                )
                .execute()
            )
            busy = response.get("calendars", {}).get("primary", {}).get("busy", [])
            return not busy
        except Exception as exc:
            logger.error("calendar is_free(%s) failed, treating as free: %s", email, exc)
            return True

    def create_hold(self, email: str, title: str, start: datetime, end: datetime, description: str = "") -> None:
        self._insert(email, title, start, end, description)

    def create_reminder(self, email: str, title: str, due: datetime, description: str = "") -> None:
        self._insert(email, title, due, due + timedelta(minutes=REMINDER_MINUTES), description)

    def _insert(self, email: str, title: str, start: datetime, end: datetime, description: str) -> None:
        try:
            service = self._service(email)
            if service is None:
                logger.info("[calendar:no-consent] hold/reminder skipped -> %s | %s", email, title)
                return
            service.events().insert(
                calendarId="primary",
                body={
                    "summary": title,
                    "description": description or "",
                    "start": {"dateTime": start.isoformat(), "timeZone": TIMEZONE},
                    "end": {"dateTime": end.isoformat(), "timeZone": TIMEZONE},
                },
            ).execute()
        except Exception as exc:
            logger.error("calendar insert(%s, %r) failed: %s", email, title, exc)


_cached = None


def calendar_service():
    """The configured calendar backend. Chosen once, then reused."""
    global _cached
    if _cached is not None:
        return _cached

    if not getattr(settings, "CALENDAR_ENABLED", False):
        _cached = StubCalendar()
        return _cached

    _cached = GoogleCalendar()
    logger.info("calendar: using per-member Google Calendar (OAuth consent)")
    return _cached


def reset_cache() -> None:
    """Test hook — forces the next call to re-read configuration."""
    global _cached
    _cached = None
